Be/keyword arguments (#27)
gSchool/ada-1-intro-to-ruby: 02-methods/10-keyword-arguments.md

# Keyword and Optional Arguments

## Learning Goals

> - Differentiate between _positional_ and _keyword_ arguments

Reviewer: Minor formatting just keeping the pattern Simon had introduced.

```suggestion

By the end of this lesson, we will be able to...

- Differentiate between _positional_ and _keyword_ arguments
```

    price = 3.7
  when :cappuccino
    price = 3.2
  else
    puts "Invalid coffee type: #{type}"
    return
  end

  # Modify for size
  case size
  when :tall
    # No change
  when :grande
    price *= 1.3
  when :venti
    price *= 1.6
  else
    puts "Invalid size: #{size}"
    return
  end

  return price
end
```

The `coffee_price` method would be invoked as follows:

```ruby
price = coffee_price(:latte, :venti)
puts "Your total is #{price} dollars"
```

Spend some time reading through the above code, then answer the following questions:

<!-- >>>>>>>>>>>>>>>>>>>>>> BEGIN CHALLENGE >>>>>>>>>>>>>>>>>>>>>> -->
<!-- Replace everything in square brackets [] and remove brackets  -->

### !challenge

* type: short-answer
* id: 46bfaed5-da9b-48b4-a087-a4303ddd9fef
* title: Method Signature
<!-- * points: [1] (optional, the number of points for scoring as a checkpoint) -->
<!-- * topics: [python, pandas] (optional the topics for analyzing points) -->

##### !question

- What's the _method signature_ of the `coffee_price` method?

##### !end-question

##### !placeholder

answer

##### !end-placeholder

##### !answer

/coffee_price\s?(type,\s?size\s?)/

##### !end-answer

> 

Reviewer: I suggest adding an explanation or hint just to make it clear to students.

We're going to tackle _#3 The caller must provide values for all arguments, even for the "standard" way of doing things (e.g. no extra shots)_ first by introducing ruby's support of **Optional Arguments**.

Optional arguments enable you to set default values when defining a function so that the caller doesn't need to pass them unless the value varies from the default. 

```ruby
# Define the method
def coffee_price(type, size, extra_shots = 0, cold = false)
  # ... calculate price, same code as before ...
  return price
end

# Price for a venti latte with 3 extra shots
coffee_price(:latte, :venti, 3)

# Price for a tall iced drip
coffee_price(:drip, :tall, 0, true)

# Price for a grande cappuccino
coffee_price(:cappuccino, :grande)
```

Notice that not all the arguments in the method above are optional. Optional Arguments can co-exist with Positional Arguments (and often do).

Positional arguments must come first, both in the definition and when the method is called.

Optional arguments can help make method invocations easier to read and write, and you'll see them frequently in Ruby, particularly when working with gems or other peoples' code. However, the following issues:

> 1. The caller must remember the order of the arguments
> 1. When reading the call, there's no way to tell which argument is which

Reviewer 2: ```suggestion
1. The caller must remember the order of the arguments
2. When reading the call, there's no way to tell which argument is which
```

are still unaddressed.

### Keyword Arguments

We can address these issues with a technique called _keyword arguments_. Here's what they look like:

```ruby
# Define the method
def coffee_price(type, size, extra_shots:, cold:)
  # ... calculate price, same code as before ...
  return price
end

# Price for a venti latte with 3 extra shots
coffee_price(:latte, :venti, extra_shots: 3, cold: false)

# Price for a tall iced drip
coffee_price(:drip, :tall, extra_shots: 0, cold: true)

# Price for a grande cappuccino
coffee_price(:cappuccino, :grande, extra_shots: 0, cold: false)
```

There are a few things to note here:
- When a method is invoked, any keyword arguments are passed by name (a.k.a. _keyword_)

> - Keyword arguments can be passed in any order. (ie. `coffee_price(:drip, :tall, cold: true, extra_shots:2`))

Reviewer 2: ```suggestion
- Keyword arguments can be passed in any order. (ie. `coffee_price(:drip, :tall, cold: true, extra_shots:2)`)
```

- Keyword Arguments can also co-exist with Positional Arguments
  - Positional arguments must come first, both in the definition and when the method is called

So new we've learned how to solve issue #3 and issues #1 & #2, but we haven't solved them all at once.

### Optional Keyword Arguments

Why not utilize both concepts? We can also make keyword arguments optional by adding a default value after the colon. 

```ruby
# Define the method
def coffee_price(type, size, extra_shots: 0, cold: false)
  # ... calculate price, same code as before ...
  return price
end

# Price for a venti latte with 3 extra shots
coffee_price(:latte, :venti, extra_shots: 3, cold: false)

# Price for a tall iced drip
coffee_price(:drip, :tall, cold: true)

# Price for a grande cappuccino
coffee_price(:cappuccino, :grande)
```

> In each of these iterations, the only things we've changed are the parameters in the method signature, the body of the method remains the same. The end result should resemble [this code](source/CoffeePrice.rb).

Reviewer 2: It looks like this PR doesn't include this file.

Reviewer 2: ```suggestion
In each of these iterations, the only things we've changed are the parameters in the method signature, the body of the method remains the same. 

<details>
  <summary>Expand this to see the end result</summary>
  
  <pre lang="ruby">
def coffee_price(type, size, extra_shots: 0, cold: false)
  # Find the base price for this drink
  case type
  when :drip
    price = 1.5
  when :latte
    price = 3.7
  when :cappuccino
    price = 3.2
  else
    puts "Invalid coffee type: #{type}"
    return
  end
  
  # Modify for size
  case size
  when :tall
    # No change
  when :grande
    price *= 1.3
  when :venti
    price *= 1.6
  else
    puts "Invalid size: #{size}"
    return
  end
  # Charge for extra shots
  price += extra_shots * 0.5
  # Charge for a cold drink
  if cold
    price += 1
  end

  return price
end
  </pre>
</details>
```

Author: added code in details tag

- Sometimes a method doesn't make sense without one of the arguments
  - How can you price coffee if you don't know the size?

**Keyword arguments** are mostly useful when the number of arguments becomes more than 2 or 3.

When you make a method, and you write code to call it, ask yourself, "is it hard remember or understand which argument is which?" or conversely "is it more trouble than it's worth to type out the keywords?"

## Summary

|                | Required                                    | Optional                                  |
|----------------|---------------------------------------------|-------------------------------------------|
| **Positional** | coffee_price(type, size, extra_shots, cold) | coffee_price(type, size, extra_shots = 0, cold = false) |
| **Keyword**    | coffee_price(type:, size:, extra_shots:, cold:) | coffee_price(type, size, extra_shots: 0, cold: false) |

- **Positional arguments** are the type of method arguments we've seen so far
- **Keyword arguments** 
  - are given by name
  - can make calling a method easier when there are several arguments
- **Optional arguments** 
  - default to a specified value if not given by the caller

>   - both Positional and Keyword Arguments can be made Optional

Reviewer 2: ```suggestion
  - both _Positional Arguments_ and _Keyword Arguments_ can be made _Optional_
```

> - Keyword Arguments and Optional Arguments can co-exist with Positional Arguments

Reviewer 2: ```suggestion
- _Keyword Arguments_ and _Optional Arguments_ can co-exist with _Positional Arguments_
```
